src/nordsieck_utils.jl

```diff
@@ -186,3 +222,161 @@ function nordsieck_rewind!(cache)
+  return nothing
+end
+
+function nordsieck_order_change(cache::T, dorder) where T
+  isconstcache = T <: OrdinaryDiffEqConstantCache
+  isconstcache || ( cache = cache.const_cache )
+  @unpack order, tau = cache
+  # WIP: uncomment when finished
+  #@inbound begin
+  begin
+    # Adams order increase
+    if dorder == 1
+      if isconstcache
+        cache.z[order+2] = zero(cache.z[order+2])
+      else
+        cache.z[order+2] .= 0
+      end
+    else
+      # Adams order decrease
+      # One needs to rescale the Nordsieck vector on an order decrease
+      cache.l .= 0
+      cache.l[2] = 1
+      dt = tau[1]
+      hsum = zero(eltype(cache.tau))
+      for j in 2:order-1
+        hsum += cache.tau[j]
+        # TODO: `hscale`?
+        ξ = hsum / dt
+        for i in j:-1:1
+          cache.l[i+1] = cache.l[i+1] * ξ + cache.l[i]
+        end # for i
+      end # for j
+
+      for j in 2:order-1
+        cache.l[j+1] = order * cache.l[j] / j
+      end
+      for j in 3:order
+        # cache.z[j] = -cache.l[j] * cache.z[order+1] + cache.z[j]
+        if isconstcache
+          cache.z[j] = muladd.(-cache.l[j], cache.z[order+1], cache.z[j])
+        else
+          @. cache.z[j] = muladd(-cache.l[j], cache.z[order+1], cache.z[j])
+        end
+      end # for j
+    end # else
+  end # @inbound
+end
+
+# `η` is `dtₙ₊₁/dtₙ`
+function choose_η!(integrator, cache::T) where T
+  isconstcache = T <: OrdinaryDiffEqConstantCache
+  isconstcache || ( cache = cache.const_cache )
+  isvarorder = nordsieck_change_order(cache)
+  order = get_current_adaptive_order(integrator.alg, integrator.cache)
+  L = order + 1
+  ηq = stepsize_η!(integrator, cache, order)
+  # If the error test fails
+  if integrator.EEst >= 1
+    nordsieck_rewind!(integrator.cache)
+    cache.η = max(integrator.opts.qmin, cache.η, integrator.opts.dtmin/abs(integrator.dt))
+    return cache.η
+  end
+  # Consider change the order
+  if isvarorder
+    cache.n_wait = 2
+    ηqm1 = stepsize_η₋₁!(integrator, cache, order)
+    ηqp1 = stepsize_η₊₁!(integrator, cache, order)
+    η = max(ηqm1, ηqp1, cache.η)
+  else
+    η = ηq
+    cache.η = η
+  end
+  ( η <= integrator.opts.qsteady_max ) && ( cache.η = 1 ; return cache.η )
+  if isvarorder
+    if η == cache.η
+      cache.nextorder = order
+    elseif η == cache.η₋₁
+      cache.η = cache.η₋₁
+      cache.nextorder = order - 1
+      cache.n_wait = L
+      nordsieck_order_change(cache, -1)
+    else
+      cache.η = cache.η₊₁
+      cache.nextorder = order + 1
+      # TODO: BDF needs a different handler
+      cache.n_wait = L
+      nordsieck_order_change(cache, 1)
+    end
+  end
+  ( integrator.iter == 1 || integrator.u_modified ) && return ( cache.η = min(1e5, cache.η) )
+  cache.η = min(integrator.opts.qmax, max(integrator.opts.qmin, cache.η))
+  return cache.η
+end
+
+function stepsize_η!(integrator, cache::T, order) where T
+  L = order+1
+  cache.η = inv( (BIAS2*integrator.EEst)^inv(L) + ADDON )
```

Comment:
I don't get these biases. Are they from Sundials?

Reply:
Yes, they are in SUNDIALS. https://github.com/LLNL/sundials/blob/master/src/cvode/cvode.c#L3036

Reply:
I see. It might just be fine-tuning of the order selection algorithm.